File: CMPS12B: Introduction to Data Structures/Lab 7: Dictionary Java/Dictionary.java
/** ***************************************************************************
* Name: Juan D. Cardozo
* Student I.D: 1362585
* CRUZ I.D: jcardozo
* Class: CMPS 12M
* Section: Mondays 6-8
* Date: 05/27/2015
* Dictionary.java
*
*****************************************************************************/

public class Dictionary implements DictionaryInterface{

	//----------------- Private---------------------

	//Node Class
	private class Node{
		String key;
		String value;
		Node left;
		Node right;
		
		//Node Constructor
		public Node(String k, String v){
			key = k;
			value = v;
			left = null;
			right = null;
		}
	}
	
	//----------------Private Dictionary Field----------
	private int numPairs;
	private Node root;
	
	//----------------Private Dictioanry Methods----------
	
	// findKey()
	// returns the Node containing key k in the subtree rooted at R, or returns
	// NULL if no such Node exists
        Node findKey(Node R ,String k){
		   if(R == null || k.compareTo(R.key)==0){ 
		      return R;
		   }else if( k.compareTo(R.key)<0 ) {
		       return (findKey(R.left, k));
		   }else{ // compareTo(R.right)
		       return (findKey(R.right, k));
		   }
	}
	
	// findParent()
	// returns the parent of N in the subtree rooted at R, or returns NULL 
	// if N is equal to R. (pre: R != NULL)
	Node findParent(Node N, Node R){
			String temp = "";
		   Node P=null;
		   if( N!=R ){
		      P = R;
		      while( P.left!=N && P.right!=N ){
		    	 temp = N.key;
		         if(temp.compareTo(P.key)<0)
		            P = P.left;
		         else
		            P = P.right;
		      }
		   }
		   return P;
		}
	
	// findLeftmost()
	// returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL.
	Node findLeftmost(Node R){
		   Node L = R;
		   if( L!=null ) for( ; L.left!=null; L=L.left) ;
		   return L;
		}
	
	// printInOrder()
	// prints the (key, value) pairs belonging to the subtree rooted at R in order
	// of increasing keys to file pointed to by out.
	String  printInOrder(Node R){
		String s ="";
	   if( R!=null){
	      s += printInOrder(R.left);
	      s += R.key;
	      s += " ";
	      s += R.value;
	      s += "\n";
	      s+=printInOrder(R.right);
	   }
	   return s;
	}
	
	//---------------Public Dictionary -----------------
	
	// Dictionary Constructor
	public Dictionary(){
		numPairs = 0;
		root = null;
	}
	
	// isEmpty()
	   // returns true if this Dictionary is empty, false otherwise
	   // pre: none
	   public boolean isEmpty(){
		   return (numPairs == 0);
	   }

	   // size()
	   // returns the number of entries in this Dictionary
	   // pre: none
	   public int size(){
		   return numPairs;
	   }

	   // lookup()
	   // returns value associated key, or null reference if no such key exists
	   // pre: none
	   public String lookup(String key){
		   Node N;
		   N = findKey(root,key);
		   return (N == null ? null : N.value);
	   }

	   // insert()
	   // inserts new (key,value) pair into this Dictionary
	   // pre: lookup(key)==null
	   public void insert(String key, String value) throws KeyCollisionException{
		  Node N,A,B;
		  if(findKey(root,key)!=null){
		   throw new KeyCollisionException("Dictionary Error: cannot insert() duplicate key:"+key);
	   }
	   N = new Node(key,value);
	   B = null;
	   A = root;
	   while(A != null){
		   B = A;
		   if(key.compareTo(A.key) < 0){ 
			   A = A.left;
		   }else{
			   A = A.right;
		   }
	   }
	   if( B == null){
		   root = N;
	   }else if(key.compareTo(B.key) < 0){
		   B.left = N;
	   }else{
		   B.right = N;
	   }
	   numPairs++;
	   }

	   // delete()
	   // deletes pair with the given key
	   // pre: lookup(key)!=null
	   public void delete(String key) throws KeyNotFoundException{
	       Node N, S, P;
	       N = findKey(root,key);
		   if(N==null){
			   throw new KeyNotFoundException("Dictionary Error: cannot delete() non-existent key:"+key);
		   }
		  
		   if( N.left==null && N.right==null ){
		      if( N == root ){
		         root = null;
		      }else{
		         P = findParent(N, root);
		         if( P.right==N ){
		        	 P.right = null;
		         }else{
		        	 P.left = null;
		         }
		      }
		   }else if( N.right==null ){
		       if( N== root ){
		         root = N.left;
		         }else{
		        	 P = findParent(N, root);
		        	 if( P.right==N ){
		        		 P.right = N.left;
		        	 }else{
		        		 P.left = N.left;
		        	 }
		         }
		       }else if( N.left==null ){
			      if( N==root ){
			         root = N.right;
			      }else{
			         P = findParent(N,root);
			         if( P.right==N ){
			        	 P.right = N.right;
			         }else{
			        	 P.left = N.right;
			         }
			      }
			   }else{  // N->left!=NULL && N->right!=NULL
			      S = findLeftmost(N.right);
			      N.key = S.key;
			      N.value = S.value;
			      P = findParent(S, N);
			      if( P.right==S ){
			    	  P.right = S.right;
			      }else{
			    	  P.left = S.right;
			      }
			   }
		   numPairs--;
	   }

	   // makeEmpty()
	   // pre: none
	   public void makeEmpty(){
		   root = null;
		   numPairs = 0;
	   }

	   // toString()
	   // returns a String representation of this Dictionary
	   // overrides Object's toString() method
	   // pre: none
	   public String toString(){
		  return  printInOrder(root);
	   }
	
}

File: CMPS12B: Introduction to Data Structures/Lab 5: Dictionary C/Dictionary.c
/** ***************************************************************************
 * Name: Juan D. Cardozo
 * Student I.D: 1362585
 * CRUZ I.D: jcardozo
 * Class: CMPS 12M
 * Section: Mondays 6-8
 * Date: 05/09/2015
 * Dictionary.c
 * It implements Dictionary.h
 ******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj

typedef struct NodeObj {
    char* key;
    char* value;
    struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type

Node newNode(char* x, char* y)
{
    Node N = malloc(sizeof (NodeObj));
    assert(N != NULL);
    N->key = x;
    N->value = y;
    N->next = NULL;
    return (N);
}

// freeNode()
// destructor for the Node type

void freeNode(Node* pN)
{
    if (pN != NULL && *pN != NULL) {
        free(*pN);
        *pN = NULL;
    }
}

// Dictionary

typedef struct DictionaryObj {
    Node top;
    int numItems;
}DictionaryObj;


// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type

Dictionary newDictionary(void)
{
    Dictionary D = malloc(sizeof (DictionaryObj));
    assert(D != NULL);
    D->top = NULL;
    D->numItems = 0;
    return D;
}

// freeDictionary()
// destructor for the Stack type

void freeDictionary(Dictionary* pD)
{
    if (pD != NULL && *pD != NULL) {
        if (!isEmpty(*pD)) makeEmpty(*pD);
        free(*pD);
        *pD = NULL;
    }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none

int isEmpty(Dictionary D)
{
    if (D == NULL) {
        fprintf(stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    if (D->numItems == 0) {
        return 1;
    } else {
        return 0;
    }
}

//size()

int size(Dictionary D)
{
    return D->numItems;
}

// lookup()
//returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none

char* lookup(Dictionary D, char* k)
{
  Node N = D->top;
  if(strcmp(N->key,k) == 0){
    return N->value;
  }else{
    while (N->next != NULL) {

        if (strcmp(k, N->key) == 0) {
            return N->value;
        }
        N = N->next;
    }

    return NULL;
  }
}

// insert()
// inserts key and value into D
// pre: none

void insert(Dictionary D, char* x, char* y)
{
    Node N;
    if (D == NULL) {
        fprintf(stderr, "Dictionary Error: calling insert() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }else if(D->numItems == 0){
      N = newNode(x, y);
      N->next = D->top;
      D->top = N;
    }else{
      N = D->top;
      while(N->next != NULL){
	N = N->next;
      }
      N->next = newNode(x,y);
    }
    D->numItems++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL

void delete(Dictionary D, char* k)
{
    Node P;
    Node garbageNode;
    if (D == NULL) {
        fprintf(stderr, "Dictionary Error: calling delete() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    P = D->top;
    if(strcmp(P->key,k) == 0){
      garbageNode = P;
      D->top = P->next;
      D->numItems--;
    }else{
    while (strcmp(k, P->next->key) != 0) {
   	P = P->next;
    }
    garbageNode = P->next;
    P->next = P->next->next;
    D->numItems--;
    }
    freeNode(&garbageNode);

}



// makeEmpty()
// re-sets D to the empty state.
// pre: none

void makeEmpty(Dictionary D)
{
    if(D == NULL){
        fprintf(stderr, "Dictionary Error: calling makeEmpty() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    Node G;
    Node N = D->top;
    while (N->next != NULL) {
      G = N;
      N = N->next;
        freeNode(&G);
    }
    G = N;
    freeNode(&G);
    D->numItems = 0;
    D->top = NULL;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out

void printDictionary(FILE* out, Dictionary D)
{
    Node N;
    if (D == NULL) {
        fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    
    for (N = D->top; N != NULL; N = N->next) fprintf(out, "key=\"%s\" value=\"%s\"\n", N->key, N->value);
    fprintf(out,"\n");
}
